Documentation/bpf.rst

```diff
@@ -2964,13 +2964,266 @@ Program Types
+cases.
+
+* **DDoS mitigation, firewalling**
+
+  One of the basic XDP BPF features is to tell the driver to drop a packet
+  with ``XDP_DROP`` at this early stage which allows for any kind of efficient
+  network policy enforcement with having an extremely low per packet cost.
+  This is ideal in situations when needing to cope with any sort of DDoS
+  attacks, but also more general allows to implement any sort of firewalling
+  policies with close to no overhead in BPF e.g. in either case as stand alone
+  appliance (e.g. scrubbing 'clean' traffic through ``XDP_TX``) or widely
+  deployed on nodes protecting end hosts themselves (via ``XDP_PASS`` or
+  cpumap ``XDP_REDIRECT`` for good traffic). Offloaded XDP takes this even
+  one step further by moving the already small per packet cost entirely
+  into the NIC with processing at line-rate.
+
+..
+
+* **Forwarding and load-balancing**
+
+  Another major use case of XDP is packet forwarding and load-balancing
+  through either ``XDP_TX`` or ``XDP_REDIRECT`` actions. The packet can
+  be arbitrarily mangled by the BPF program running in the XDP layer,
+  even BPF helper functions are available for increasing or decreasing
+  the packet's headroom in order to arbitrarily encapsulate respectively
+  decapsulate the packet before sending it out again. With ``XDP_TX``
+  one-legged load-balancers can be implemented that push the packet out
+  of the same networking device it originally arrived on, or with the
+  ``XDP_REDIRECT`` action it can be forwarded to another NIC for
+  transmission. The latter return code can also be used in combination
+  with BPF's cpumap to load-balance packets for passing up the local
+  stack, but on remote, non-XDP processing CPUs.
+
+..
+
+* **Pre-stack filtering / processing**
+
+  Besides policy enforcement, XDP can also be used for hardening the
+  kernel's networking stack with the help of ``XDP_DROP`` case, meaning,
+  it can drop irrelevant packets for a local node right at the earliest
+  possible point before the networking stack sees them e.g. given we
+  know that a node only serves TCP traffic, any UDP, SCTP or other L4
+  traffic can be dropped right away. This has the advantage that packets
+  do not need to traverse various entities like GRO engine, the kernel's
+  flow dissector and others before it can be determined to drop them and
+  thus this allows for reducing the kernel's attack surface. Thanks to
+  XDP's early processing stage, this effectively 'pretends' to the kernel's
+  networking stack that these packets have never been seen by the networking
+  device. Additionally, if a potential bug in the stack's receive path
+  got uncovered and would cause a 'ping of death' like scenario, XDP can be
+  utilized to drop such packets right away without having to reboot the
+  kernel or restart any services, meaning, XDP can just atomically swap
+  such drop enforcement of bad packets without any network traffic
+  interruption on an end host.
```

Comment:
the last half of this sentence, from "meaning, ..." doesn't seem to add to the point of allowing to drop such packets right away. This atomic property is also described earlier, so I would suggest either dropping it, or stopping the prior sentence and starting a new sentence to describe the benefits of being able to swap these "drop harmful traffic" programs on-the-fly, and why this is interesting to the pre-stack filtering case.